app/data_sources/gitlab.py

```diff
@@ -0,0 +1,107 @@
+
+
+class GitlabDataSource(BaseDataSource):
+    @staticmethod
+    def validate_config(config: Dict) -> None:
+        try:
+            parsed_config = GitlabConfig(**config)
+            session = requests.Session()
+            session.headers.update({"PRIVATE-TOKEN": parsed_config.access_token})
+            projects_response = session.get(PROJECTS_URL)
+            if projects_response.status_code != 200:
+                raise ValueError("Invalid api key")
+        except (KeyError, ValueError) as e:
+            raise InvalidDataSourceConfig from e
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # Create a access token with sufficient permissions in https://gitlab.com/-/profile/personal_access_tokens
+        self.gitlab_config = GitlabConfig(**self._config)
+        self.session = requests.Session()
+        self.session.headers.update({"PRIVATE-TOKEN": self.gitlab_config.access_token})
+
+    def _feed_new_documents(self) -> None:
+        projects_response = self.session.get(PROJECTS_URL)
+        projects = projects_response.json()
+
+        self._parse_projects_in_parallel(projects)
+
+    def _parse_projects_worker(self, projects):
+
+        documents = []
+
+        for project in projects:
+            project_id = project["id"]
+            issues_url = f"{GITLAB_BASE_URL}/projects/{project_id}/issues"
+            issues_response = self.session.get(issues_url)
+            issues_json = issues_response.json()
+
+            for issue in issues_json:
+                last_modified = datetime.strptime(issue["updated_at"], "%Y-%m-%dT%H:%M:%S.%fZ")
+                if last_modified < self._last_index_time:
+                    continue
+
+                documents.append(BasicDocument(
+                    id=issue["id"],
+                    data_source_id=self._data_source_id,
+                    type=DocumentType.DOCUMENT,
+                    title=issue['title'],
+                    content=issue["description"] if not None else "",
+                    author=issue['author']['name'],
+                    author_image_url=issue['author']['avatar_url'],
+                    location=project["web_url"],
+                    url=issue['web_url'],
+                    timestamp=last_modified
+                ))
+
+            pull_requests_url = f"{GITLAB_BASE_URL}/projects/{project_id}/merge_requests"
+            pull_requests_response = self.session.get(pull_requests_url)
+            pull_requests_json = pull_requests_response.json()
+
+            for pull_request in pull_requests_json:
+                last_modified = datetime.strptime(pull_request["updated_at"], "%Y-%m-%dT%H:%M:%S.%fZ")
+                if last_modified < self._last_index_time:
+                    continue
+
+                documents.append(BasicDocument(
+                    id=pull_request["id"],
+                    data_source_id=self._data_source_id,
+                    type=DocumentType.DOCUMENT,
```

Comment:
Change to GIT_PR (pull first, i'm adding it now)

Reply:
Done